### PR TITLE
[BREAKING] Change type of Super to TypeType

### PR DESCRIPTION
**Pull request by hrkrshnn**
Closes https://github.com/ethereum/solidity/issues/7744

### CODE REVIEW

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `ed5768bb9cbd7cab2aac390903d4f985a41068ae`:
```
Coding style error:
 libsolidity/analysis/TypeChecker.cpp:2768: else if(dynamic_cast<ContractType const*>(exprType))
 libsolidity/codegen/ExpressionCompiler.cpp:1419: _memberAccess.expression().accept(*this);

```
Please check that your changes are working as intended.

File: libsolidity/analysis/TypeChecker.cpp [hunk 1]
```diff
@@ -2756,8 +2756,18 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
...
+		{
+			if (auto typeType = dynamic_cast<TypeType const*>(exprType))
+			{
+				if (
+					auto contractType = dynamic_cast<ContractType const*>(typeType->actualType());
+					contractType->isSuper()
+				)
+					requiredLookup = VirtualLookup::Super;
+			}
+			else if(dynamic_cast<ContractType const*>(exprType))
```
Comment (stackenbotten) [inline libsolidity/analysis/TypeChecker.cpp:2768]: Coding style error

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 2]
```diff
@@ -1347,64 +1347,77 @@ bool ExpressionCompiler::visit(MemberAccess const& _memberAccess)
...
+					default:
+						solAssert(false, "unsupported member function");
+					}
 				}
+				else if (dynamic_cast<TypeType const*>(_memberAccess.annotation().type))
+				{
+					// no-op
+				}
+				else
+ 					_memberAccess.expression().accept(*this);
```
Comment (stackenbotten) [inline libsolidity/codegen/ExpressionCompiler.cpp:1419]: Coding style error

File: libsolidity/analysis/TypeChecker.cpp [hunk 3]
```diff
@@ -2756,8 +2756,18 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
...
-			if (auto contractType = dynamic_cast<ContractType const*>(exprType))
-				requiredLookup = contractType->isSuper() ? VirtualLookup::Super : VirtualLookup::Virtual;
+		{
+			if (auto typeType = dynamic_cast<TypeType const*>(exprType))
+			{
+				if (
+					auto contractType = dynamic_cast<ContractType const*>(typeType->actualType());
+					contractType->isSuper()
+				)
+					requiredLookup = VirtualLookup::Super;
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: Are you sure it always is a contract type?
```suggestion
				auto contractType = dynamic_cast<ContractType const*>(typeType->actualType());
				if (contractType && contractType->isSuper())
					requiredLookup = VirtualLookup::Super;
```

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 4]
```diff
@@ -1967,7 +1976,7 @@ void ExpressionCompiler::endVisit(Identifier const& _identifier)
 		{
 		case Type::Category::Contract:
 			// "this" or "super"
```
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp]: Is the comment still correct?

Comment (hrkrshnn) [reply libsolidity/codegen/ExpressionCompiler.cpp]: I changed it and added an assert.

File: libsolidity/analysis/TypeChecker.cpp (hunk 3 above)
Comment (hrkrshnn) [reply libsolidity/analysis/TypeChecker.cpp]: Fixed.

**Comment by hrkrshnn** [comment]
The failing tests are all SMTChecker tests.

File: libsolidity/analysis/TypeChecker.cpp [hunk 5]
```diff
@@ -2750,8 +2750,16 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
 			);
 
 		if (!funType->bound())
-			if (auto contractType = dynamic_cast<ContractType const*>(exprType))
-				requiredLookup = contractType->isSuper() ? VirtualLookup::Super : VirtualLookup::Virtual;
+		{
+			if (auto typeType = dynamic_cast<TypeType const*>(exprType))
```
Comment (leonardoalt) [inline libsolidity/analysis/TypeChecker.cpp:2763]:
```suggestion
			if (auto const* typeType = dynamic_cast<TypeType const*>(exprType))
```

File: libsolidity/analysis/TypeChecker.cpp [hunk 6]
```diff
@@ -2750,8 +2750,16 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
...
+			{
+				auto contractType = dynamic_cast<ContractType const*>(typeType->actualType());
```
Comment (leonardoalt) [inline libsolidity/analysis/TypeChecker.cpp:2765]:
```suggestion
				auto const* contractType = dynamic_cast<ContractType const*>(typeType->actualType());
```

File: libsolidity/analysis/TypeChecker.cpp [hunk 7]
```diff
@@ -2750,8 +2750,16 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
...
+				if (contractType && contractType->isSuper())
+					requiredLookup = VirtualLookup::Super;
+			}
+			else if (dynamic_cast<ContractType const*>(exprType))
+				requiredLookup = VirtualLookup::Virtual;
```
Comment (leonardoalt) [inline libsolidity/analysis/TypeChecker.cpp]: Could change this `else if` to `else` and assert the condition.

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 8]
```diff
@@ -1355,59 +1355,72 @@ bool ExpressionCompiler::visit(MemberAccess const& _memberAccess)
 	// for internal functions, or enum/struct definitions.
 	if (TypeType const* type = dynamic_cast<TypeType const*>(_memberAccess.expression().annotation().type))
 	{
-		if (dynamic_cast<ContractType const*>(type->actualType()))
+		if (auto contractType = dynamic_cast<ContractType const*>(type->actualType()))
```
Comment (leonardoalt) [inline libsolidity/codegen/ExpressionCompiler.cpp:1358]:
```suggestion
		if (auto const* contractType = dynamic_cast<ContractType const*>(type->actualType()))
```

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 9]
```diff
@@ -1355,59 +1355,72 @@ bool ExpressionCompiler::visit(MemberAccess const& _memberAccess)
...
 		{
 			solAssert(_memberAccess.annotation().type, "_memberAccess has no type");
-			if (auto variable = dynamic_cast<VariableDeclaration const*>(_memberAccess.annotation().referencedDeclaration))
-				appendVariable(*variable, static_cast<Expression const&>(_memberAccess));
-			else if (auto funType = dynamic_cast<FunctionType const*>(_memberAccess.annotation().type))
+			if (contractType->isSuper())
 			{
-				switch (funType->kind())
+				_memberAccess.expression().accept(*this);
+				solAssert(!!_memberAccess.annotation().referencedDeclaration, "Referenced declaration not resolved.");
```
Comment (leonardoalt) [inline libsolidity/codegen/ExpressionCompiler.cpp]:
```suggestion
				solAssert(_memberAccess.annotation().referencedDeclaration, "Referenced declaration not resolved.");
```
I don't really like the `!!` style, so commenting that. Fine if you want to keep

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 10]
```diff
@@ -1593,7 +1597,9 @@ bool ExpressionCompiler::visit(MemberAccess const& _memberAccess)
 		else if (member == "creationCode" || member == "runtimeCode")
 		{
 			TypePointer arg = dynamic_cast<MagicType const&>(*_memberAccess.expression().annotation().type).typeArgument();
-			ContractDefinition const& contract = dynamic_cast<ContractType const&>(*arg).contractDefinition();
+			auto contractType = dynamic_cast<ContractType const*>(arg);
```
Comment (leonardoalt) [inline libsolidity/codegen/ExpressionCompiler.cpp]:
```suggestion
			auto const& contractType = dynamic_cast<ContractType const&>(*arg);
```

Comment (leonardoalt) [reply libsolidity/codegen/ExpressionCompiler.cpp]: Pointer is being used without checking for nullptr, so suggesting to change to reference

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 11]
```diff
@@ -1610,7 +1616,10 @@ bool ExpressionCompiler::visit(MemberAccess const& _memberAccess)
 		else if (member == "name")
 		{
 			TypePointer arg = dynamic_cast<MagicType const&>(*_memberAccess.expression().annotation().type).typeArgument();
-			ContractDefinition const& contract = dynamic_cast<ContractType const&>(*arg).contractDefinition();
+			auto contractType = dynamic_cast<ContractType const*>(arg);
```
Comment (leonardoalt) [inline libsolidity/codegen/ExpressionCompiler.cpp]:
```suggestion
			auto const& contractType = dynamic_cast<ContractType const&>(*arg);
```

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 12]
```diff
@@ -884,17 +884,17 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
 				solAssert(*identifier->annotation().requiredLookup == VirtualLookup::Virtual, "");
 				functionDef = &functionDef->resolveVirtual(m_context.mostDerivedContract());
 			}
-			else
-			{
-				ContractType const* type = dynamic_cast<ContractType const*>(memberAccess->expression().annotation().type);
-				if (type && type->isSuper())
+			else if (auto typeType = dynamic_cast<TypeType const*>(memberAccess->expression().annotation().type))
```
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:887]:
```suggestion
			else if (auto const* typeType = dynamic_cast<TypeType const*>(memberAccess->expression().annotation().type))
```

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 13]
```diff
@@ -1649,7 +1638,9 @@ void IRGeneratorForStatements::endVisit(MemberAccess const& _memberAccess)
 		else if (member == "creationCode" || member == "runtimeCode")
 		{
 			TypePointer arg = dynamic_cast<MagicType const&>(*_memberAccess.expression().annotation().type).typeArgument();
-			ContractDefinition const& contract = dynamic_cast<ContractType const&>(*arg).contractDefinition();
+			auto contractType = dynamic_cast<ContractType const*>(arg);
```
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]:
```suggestion
			auto const& contractType = dynamic_cast<ContractType const&>(*arg);
```

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 14]
```diff
@@ -1669,7 +1660,9 @@ void IRGeneratorForStatements::endVisit(MemberAccess const& _memberAccess)
 		else if (member == "interfaceId")
 		{
 			TypePointer arg = dynamic_cast<MagicType const&>(*_memberAccess.expression().annotation().type).typeArgument();
-			ContractDefinition const& contract = dynamic_cast<ContractType const&>(*arg).contractDefinition();
+			auto contractType = dynamic_cast<ContractType const*>(arg);
```
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]:
```suggestion
			auto const& contractType = dynamic_cast<ContractType const&>(*arg);
```

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 15]
```diff
@@ -2095,18 +2107,20 @@ void IRGeneratorForStatements::endVisit(Identifier const& _identifier)
...
-			}
+			solAssert(_identifier.name() == "this", "");
+			define(_identifier) << "address()\n";
 			break;
 		case Type::Category::Integer:
 			solAssert(_identifier.name() == "now", "");
 			define(_identifier) << "timestamp()\n";
 			break;
+		case Type::Category::TypeType:
+			{
```
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Indentation one level <<

File: libsolidity/analysis/TypeChecker.cpp (hunk 5 above)
Comment (hrkrshnn) [reply libsolidity/analysis/TypeChecker.cpp:2763]: The deduced type is the same in both cases. In fact the conversions `T const*` into `T*` and `T const*` into `T` are type errors, so the deduced type should be quite clear here. I would prefer not writing `const *` twice, once inside `dynamic_cast` and the other next to auto.

Comment (hrkrshnn) [reply libsolidity/analysis/TypeChecker.cpp:2763]: Although in the codebase, both forms are used,  both `auto x = dynamic_cast<T const*>(...)` and `auto const* = dynamic_cast<T const*>(...)`

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 9 above)
Comment (hrkrshnn) [reply libsolidity/codegen/ExpressionCompiler.cpp]: I don't like it either. This was a copy-paste. I can remove it.

File: libsolidity/analysis/TypeChecker.cpp (hunk 5 above)
Comment (leonardoalt) [reply libsolidity/analysis/TypeChecker.cpp:2763]: Yea, I usually use `auto const*` to help/force some less clever compilers, but fine with either

**Comment by chriseth** [comment]
Tests are failing.

File: libsolidity/analysis/TypeChecker.cpp (hunk 7 above)
Comment (hrkrshnn) [reply libsolidity/analysis/TypeChecker.cpp]: Apparently test fails if you do this.

Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp]: Actually isn't "virtual" wrong here?

File: libsolidity/analysis/ViewPureChecker.cpp [hunk 16]
```diff
@@ -195,7 +195,7 @@ void ViewPureChecker::endVisit(Identifier const& _identifier)
 		{
 		case Type::Category::Contract:
 			solAssert(_identifier.name() == "this" || _identifier.name() == "super", "");
```
Comment (chriseth) [inline libsolidity/analysis/ViewPureChecker.cpp]:
```suggestion
			solAssert(_identifier.name() == "this", "");
```

File: libsolidity/ast/Types.cpp [hunk 17]
```diff
@@ -2149,34 +2149,7 @@ MemberList::MemberMap ContractType::nativeMembers(ASTNode const*) const
 {
 	MemberList::MemberMap members;
 	if (m_super)
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp]:
```suggestion
solAssert(!m_super, "")
```

File: test/libsolidity/semanticTests/metaTypes/name_other_contract.sol [hunk 18]
```diff
@@ -11,7 +11,7 @@ contract C {
     }
 }
 
-contract Test {
+contract Test is C {
```
Comment (chriseth) [inline test/libsolidity/semanticTests/metaTypes/name_other_contract.sol:14]: Why did you add the inheritance?

Comment (chriseth) [reply test/libsolidity/semanticTests/metaTypes/name_other_contract.sol:14]: Ah, because of the super!

File: test/libsolidity/semanticTests/metaTypes/name_other_contract.sol [hunk 19]
```diff
@@ -21,9 +21,13 @@ contract Test {
     function i() public pure returns (string memory) {
         return type(I).name;
     }
+    function j() public pure returns (string memory) {
+        return type(super).name;
```
Comment (chriseth) [inline test/libsolidity/semanticTests/metaTypes/name_other_contract.sol:25]: Is this actually working for a more complicated case?

File: libsolidity/analysis/TypeChecker.cpp (hunk 7 above)
Comment (hrkrshnn) [reply libsolidity/analysis/TypeChecker.cpp]: Why?

File: test/libsolidity/semanticTests/metaTypes/super_name.sol [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+        assert(compareStrings(r[1], "A"));
+        assert(compareStrings(r[2], "B"));
+        assert(compareStrings(r[3], "C"));
+
+        return true;
+    }
+}
+
+// ----
+// f() -> true
```
Comment (hrkrshnn) [inline test/libsolidity/semanticTests/metaTypes/super_name.sol:54]: @chriseth Added a test for `type(super).name` with a complex inheritance.

File: libsolidity/analysis/TypeChecker.cpp (hunk 7 above)
Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp]: Because it is a member access and that should always be static. I think in this case, we are doing an external call, where virtual looking doesn't even make sense.

Comment (hrkrshnn) [reply libsolidity/analysis/TypeChecker.cpp]: I changed it to `Static`. This is normally triggered in expressions such as `this.f`. I'll also change it in develop.

**Comment by hrkrshnn** [comment]
Rebased.

**Review by leonardoalt — COMMENTED** [review]
Looks good to me, but it'd be good to get another review.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 21]
```diff
@@ -1790,58 +1783,77 @@ void IRGeneratorForStatements::endVisit(MemberAccess const& _memberAccess)
...
+					switch (memberFunctionType->kind())
 					{
-						define(_memberAccess) << to_string(function->id()) << "\n";
-						m_context.internalFunctionAccessed(_memberAccess, *function);
+					case FunctionType::Kind::Declaration:
+						break;
+					case FunctionType::Kind::Internal:
+						if (auto const* function = dynamic_cast<FunctionDefinition const*>(_memberAccess.annotation().referencedDeclaration))
+						{
+							define(_memberAccess) << to_string(function->id()) << "\n";
```
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: 7 levels of indentation :(

Comment (hrkrshnn) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: I reduced it to 6. There was an `else { if ... else {}}` block. I replaced it by `else if ()... else {}`.

**Comment by hrkrshnn** [comment]
Rebased, and added [minor](https://github.com/ethereum/solidity/pull/10029#discussion_r520611836) change.

**Review by chriseth — APPROVED** [review]
